src/DI/ServiceLookup/CallSiteChain.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
+{
+    internal class CallSiteChain
+    {
+        private readonly Dictionary<Type,ChainItemInfo> _callSiteChain;
+
+        public CallSiteChain()
+        {
+            _callSiteChain = new Dictionary<Type, ChainItemInfo>();
+        }
+
+        public void CheckCircularDependency(Type serviceType)
+        {
+            if (_callSiteChain.ContainsKey(serviceType))
+            {
+                throw new InvalidOperationException(CreateCircularDependencyExceptionMessage(serviceType));
+            }
+        }
+
+        public void Remove(Type serviceType)
+        {
+            _callSiteChain.Remove(serviceType);
+        }
+
+        public void Add(Type serviceType, Type implementationType = null)
+        {
+            _callSiteChain[serviceType] = new ChainItemInfo(_callSiteChain.Count, implementationType);
+        }
+
+        private string CreateCircularDependencyExceptionMessage(Type type)
+        {
+            var messageBuilder = new StringBuilder();
+            messageBuilder.AppendFormat(Resources.CircularDependencyException, TypeNameHelper.GetTypeDisplayName(type));
+            messageBuilder.AppendLine();
+
+            AppendResolutionPath(messageBuilder, type);
+
+            return messageBuilder.ToString();
+        }
+
+        private void AppendResolutionPath(StringBuilder builder, Type currentlyResolving = null)
+        {
+            foreach (var pair in _callSiteChain.OrderBy(p => p.Value.Order))
+            {
+                var serviceType = pair.Key;
+                var implementationType = pair.Value.ImplementationType;
+                if (implementationType == null || serviceType == implementationType)
+                {
+                    builder.AppendFormat(Resources.ResolutionPathServiceType, TypeNameHelper.GetTypeDisplayName(serviceType));
+                }
+                else
+                {
+                    builder.AppendFormat(Resources.ResolutionPathServiceAndImplementationType,
+                        TypeNameHelper.GetTypeDisplayName(serviceType),
+                        TypeNameHelper.GetTypeDisplayName(implementationType));
+                }
+
+                builder.Append(Resources.ResolutionPathSeparator);
+            }
+
+            builder.AppendFormat(Resources.ResolutionPathServiceType, TypeNameHelper.GetTypeDisplayName(currentlyResolving));
+        }
+
+        private struct ChainItemInfo
+        {
+            public int Order { get; }
```

Comment:
Is Order ever used? I thought the order is always a natural order due to the sequence of calls in the chain.

Reply:
Order used during resolution path display `_callSiteChain.OrderBy(p => p.Value.Order)` in `AppendResolutionPath` function.

Although `Dictionary<TKey,TValue>` currently store pair in natural order this behavior not guaranteed.

We could use `System.Collections.Specialized.OrderedDictionary`, but it's not generic and will cause a lot of boxing/unboxing. 

This could be solved if we get rid of `ChainItemInfo` and use `ImplementationType` as Value, but we won't have ability to add some additional info. For example: initially I had enum `ResolutionType`(ConstructorCall/InstanceUse/EnumerableCreate/FactoryRun) that could be used to display each case differently.

Reply:
Ok, but no one is ever setting Order as far as I can tell, so I think there's still a problem, no?

And yes, you're absolutely right that we can't depend on Dictionary returning the same order. Beyond a certain size, it's nearly guaranteed that it won't return the same order.

Reply:
Order set in `Add` method, when we create `ChainItemInfo`. It equals to count of items added in chain before, to preserve natural order.

Reply:
Oh sorry I totally missed that, you are correct. Looks good!